src/server/pps/server/api_server.go

```diff
@@ -2944,42 +2938,41 @@ func (a *apiServer) DeleteAll(ctx context.Context, request *types.Empty) (respon
 }
 
 // ActivateAuth implements the protobuf pps.ActivateAuth RPC
-func (a *apiServer) ActivateAuth(ctx context.Context, req *pps.ActivateAuthRequest) (resp *pps.ActivateAuthResponse, retErr error) {
+func (a *apiServer) ActivateAuth(ctx context.Context, req *pps.ActivateAuthRequest) (*pps.ActivateAuthResponse, error) {
+	var resp *pps.ActivateAuthResponse
+	if err := a.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		var err error
+		resp, err = a.ActivateAuthInTransaction(ctx, txnCtx, req)
+		return err
+	}); err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (a *apiServer) ActivateAuthInTransaction(ctx context.Context, txnCtx *txncontext.TransactionContext, req *pps.ActivateAuthRequest) (resp *pps.ActivateAuthResponse, retErr error) {
 	// Unauthenticated users can't create new pipelines or repos, and users can't
 	// log in while auth is in an intermediate state, so 'pipelines' is exhaustive
-	var eg errgroup.Group
-	pipelineInfo := &pps.PipelineInfo{}
-	if err := a.pipelines.ReadOnly(ctx).List(pipelineInfo, col.DefaultOptions(), func(string) error {
-		pipeline := proto.Clone(pipelineInfo).(*pps.PipelineInfo)
-		pipelineName := pipeline.Pipeline.Name
+	pi := &pps.PipelineInfo{}
+	if err := a.pipelines.ReadOnly(ctx).List(pi, col.DefaultOptions(), func(string) error {
```

Comment:
What does ReadOnly here do?  Should the list operation somehow have the txnCtx transaction passed in?  (Looks like Collection.ReadWrite takes that, and we are writing to pipelines, though not to the collection of pipelines, so I'm not 100% sure.)

Real question is what happens if a pipeline is added while this code is running.  I don't know what Postgres does under the hood; if a collection is `select * from collections.pipeline`, does inserting a row into that table invalidate the transaction?

If there's a conflict, where do we retry this transaction?  (I was guessing that things like `updatePipeline` would have a retry loop since they take a function to do the update, but I didn't see one.)

Reply:
this API is a little unusual - we don't expose List() on the ReadWrite collection because it's implemented across transactions due to buffering. This means that we are listing pipelines from outside a transaction and updating each of them inside one. This does open the possibility that a pipeline is created after the transaction starts and we fail to update it